include/h2o/socket.h

```diff
@@ -108,6 +108,11 @@ struct st_h2o_socket_t {
      * total bytes written (above the TLS layer)
      */
     size_t bytes_written;
+    /**
+     * flag to track if sock is being traced
+     */
+    int is_traced;
```

Comment:
I think it might be a good idea to change this to:
* an internal flag (prefixed by `_`)
* that is only being accessed from `h2o_socket_is_traced`
* a ternary state of (undetermined, not_traced, traced)

Users would call `h2o_socket_is_traced` (most likely via `h2o_conn_is_traced`) to see if the connection is being traced. The `h2o_socket_traced` function will consult the ternary state, and if it is in the undetermined state, check the map and update the state to either not_traced or traced. If the state is not in the undetermined state, the current state can be returned immediately.

By moving to such a design, we can get rid of the `h2o_conn_tracing` probe. The`H2O_PROBE_CONN` macro calls `_ENABLED()` before calling `h2o_conn_is_tracing`, therefore there is no need to have a special probe just to see if connection-level checks need to be made.